### PR TITLE
nv2a: Make multiplication by 0 match HW behavior.

### LINKED ISSUE #1008: America's Army: Rise of a Soldier: Floor texture Broken
### Title

https://xemu.app/titles/55530043/#America-s-Army-Rise-of-a-Soldier

### Bug Description

When going in-game the floor textures black
xemu-2022-06-01-19-23-53

### Expected Behavior

It should be green

### xemu Version

Version: 0.7.25
Branch: master
Commit: https://github.com/mborgerson/xemu/commit/7d6da227ba3b224647e35d2385419a3bec25a3b3
Date: Tue May 31 08:20:36 PM UTC 2022

### System Information

Field | Value
-- | --
OS | Windows 10
CPU | AMD Ryzen 5 2600 Six-Core Processo
Graphics Device | NVIDIA GeForce RTX 3060 Ti/PCIe/SSE2
Graphics Driver | 4.0.0 NVIDIA 512.95

### Additional Context

_No response_

### LINKED ISSUE #1044: Terminator Redemption Shadows Bug
### Title

https://xemu.app/titles/4947003e/#Terminator-3-The-Redemption

### Bug Description

Captura desde 2022-06-08 23-23-10

### Expected Behavior

shadows should display correctly

### xemu Version

0.7.29

### System Information

Manjaro Linux Release 21.2.6

AMD A10-7850K Radeon R7, 12 Compute Cores 4C+8G

mesa-22.0.4-1

Same problem with Windows 10 and Nvidia 3070M

### Additional Context

Anyway the game freezes in that scene at the beginning of the game

### PR DESCRIPTION
**Pull request by abaire**
Fixes #1008
Fixes #1044

The nv2a returns 0 for anything multiplied by zero, including exceptional
values such as Inf and NaN. Desktop GPUs do not enforce this, leading to
conditions where NaNs wipe out calculations and lead to erroneous behavior.

[Test](https://github.com/abaire/nxdk_vsh_tests/blob/main/src/tests/americasarmyshader.cpp)
[HW Results](https://github.com/abaire/nxdk_vsh_tests_golden_results/wiki/Results-AmericasArmyShader)

### CODE REVIEW

**Comment by Shoegzer** [comment]
@abaire Based on your comment in #55 about the broken floor texture in Halo 2 being related to depth buffer precision, I assumed this PR would not fix that issue.  I did test it to be sure, to no avail, in case it matters at all.

**Comment by abaire** [comment]
> @abaire Based on your comment in #55 about the broken floor texture in Halo 2 being related to depth buffer precision, I assumed this PR would not fix that issue. I did test it to be sure, to no avail, in case it matters at all.

Thanks for checking, I would've been very surprised if this fixed the issue in Halo 2, it should really only affect calculations that end up dealing with multiplication of exceptional values (Inf/NaN/etc..) which did not appear to be the case when I briefly looked at #55.

**Comment by Triticum0** [comment]
This fixes the menu issues in american chopper 1 and 2

**Comment by Triticum0** [comment]
### American Chopper
Before
Black screen
xemu-2022-06-24-13-53-50
xemu-2022-06-24-13-53-54
xemu-2022-06-24-13-53-56
After
xemu-2022-06-24-13-54-21
xemu-2022-06-24-13-54-23
xemu-2022-06-24-13-54-30
xemu-2022-06-24-13-54-32

**Comment by Triticum0** [comment]
American Chopper 2
Before
xemu-2022-06-24-13-52-20
xemu-2022-06-24-13-52-30
After
xemu-2022-06-24-13-52-03
xemu-2022-06-24-13-49-37

**Comment by Triticum0** [comment]
This pr fixes azurik - rise of perathia late game 
Before Screenshot_40

After 
Screenshot_42-1

**Comment by Triticum0** [comment]
Also FIxes crazy taxi 3 on Nivida hardware

File: hw/xbox/nv2a/vsh.c [hunk 1]
```diff
@@ -634,7 +634,16 @@ static const char* vsh_header =
     "#define MUL(dest, mask, src0, src1) dest.mask = _MUL(_in(src0), _in(src1)).mask\n"
     "vec4 _MUL(vec4 src0, vec4 src1)\n"
     "{\n"
-    "  return src0 * src1;\n"
+    // Unfortunately mix() falls victim to the same handling of exceptional
+    // (inf/NaN) handling as a multiply, so per-component comparisons are used
+    // to guarantee HW behavior (anything * 0 must == 0).
+    "  vec4 zero_components = abs(sign(src0) * sign(src1));\n"
```
Comment (PatrickvL) [inline hw/xbox/nv2a/vsh.c]: Why is there an `abs()` here if the only value checked for is `0.0` - is negative zero a case to cater for as well?

**Review by PatrickvL — COMMENTED** [review]
A genuine question

File: hw/xbox/nv2a/vsh.c (hunk 1 above)
Comment (abaire) [reply hw/xbox/nv2a/vsh.c]: Yes, the intent was to cover the potential for a -0, in practice I believe `sign` would only return 0.0 so this is probably unnecessarily conservative.

**Comment by Triticum0** [comment]
This also affects #1044 using this pr fixes the issue

**Comment by user18081972** [comment]
this PR seems to fix quite a handful of graphical issues

**Comment by Triticum0** [comment]
Also seems to fix Breakdown lighting issues https://www.youtube.com/watch?v=LQ5y4sdaA5s

**Comment by user18081972** [comment]
Maybe we should merge it

**Comment by Fabxx** [comment]
> Maybe we should merge it

this is still incomplete yet since it works only at 1x resolution scale, and we don't know if there might be edge cases where graphics might change behavior like it happened before with a previous "version" of this PR. There's no need to rush if we have to grant stability.

**Pull request merged**